Jb/cops 4829 (#2726)
mesosphere/dcos-docs-site: pages/mesosphere/dcos/services/edge-lb/1.5/concepts/edge-lb-architecture-deep-dive/index.md

---
layout: layout.pug
navigationTitle: Architectural deep-dive
title: Architectural deep-dive
menuWeight: 20
excerpt: Provides a more detailed view of Edge-LB components within the network topology for a cluster
enterprise: true
---

Edge-LB provides layer-7 inbound load balancing and proxying through the Edge-LB API server and Edge-LB pools. This section provides a deeper view into the details of the Edge-LB architecture and how it works.

The following diagram provides a more detailed representation of the Edge-LB architecture.

![Detail view of Edge-LB components](/mesosphere/dcos/services/edge-lb/1.5/img/Edge-LB-detail-arch.png)

# A closer look at the Edge-LB API server
The Edge-LB API receives incoming client requests and, based on the action requested, creates, deletes, or updates Edge-LB pool instances. The API server also serves API endpoint requests for Edge-LB pool(s) instances.

> The Edge-LB API server uses the DC/OS ZooKeeper persistent storage to store the configuration file for the load balancer and communicates to the Edge-LB pool instances through ZooKeeper&reg; transactions. The interaction between ZooKeeper and the Edge-LB API server enables Edge-LB to be fault tolerant. In the event of an API server failure, the Edge-LB pool instances continue to load-balance traffic to the proper backend. When the Edge-LB API server comes back online, it re-establishes the communication between itself and the pool instances.

Author: You can see the generated pages at:  https://docs-lisa.mesosphere.com/mesosphere/dcos/services/edge-lb/1.5/

> The Edge-LB API server uses the DC/OS ZooKeeper persistent storage to store the configuration file for the load balancer and communicates to the Edge-LB pool instances through ZooKeeper&reg; transactions. The interaction between ZooKeeper and the Edge-LB API server enables Edge-LB to be fault tolerant. In the event of an API server failure, the Edge-LB pool instances continue to load-balance traffic to the proper backend. When the Edge-LB API server comes back online, it re-establishes the communication between itself and the pool instances.

Author: Good catch! Thanks!

The Edge-LB API server has two sub-components:
- Apache&reg; Mesos&reg; listener

> - DC/OS template

Author: Created separate ticket.

## Mesos listener

> The `mesos-listener` sub-component is always up by default. It subscribes to the Mesos event stream, translates the event stream data into an internal data structure, and serves metadata from the data structure through the API server. It also runs a gRPC server to serve the Mesos task information to other Edge-LB components like the `dcos-template`.

Author: Fixed both.

## DC/OS template
The `dcos-template` sub-component dynamically renders the load balancer configuration information for HAProxy&reg; by combining the `mesos-listener` output and the Edge-LB pool configuration provided by the Edge-LB CLI client. The template then signals the `lbmgr` sub-component to take appropriate action on the HAProxy load balancer instance.

> In the event of Edge-LB pool update, the `dcos-template` re-generates the HAProxy load balancer configuration from the Edge-LB pool configuration file and signals `lbmgr` to reload the existing HAProxy configuration in real-time.

Reviewer: > and signals `lbmgr` to reload the existing HAProxy configuration in real-time.

It is not dcos-template but cloud-controller, and there is a persistent, streaming gRPC connection between cloud-controller and `lbmgr` through which the new configuration is sent. We used to do signaling but IIRC in 1.5, we switched to gRPC.

Also, the `lbmgr` in this context may confuse the reader, as it is a pretty deep implementation detail. `lbmgr` is a part of the pool server, it supervises the proxy process and coordinates reloads. Having said that, what do you think about using the term `pool server` here instead and describing the role of lbmgr later on in the text? 

I am not sure how well the term `pool server` is established in the documentation. Even within the team we have troubles using consistent naming for it. Maybe it would be nice to fix it now and agree upon`pool server`?

Author: Created separate ticket

# A closer look at Edge-LB pools
The Edge-LB pool provides the core load balancing functionality for Edge-LB software.

> However, there are no Edge-LB pool instances created by default.

Author: Fixed.

To create a pool instance, you must define configuration settings in a pool configuration file.
You can set up load balancing for one service or for multiple services using the same pool configuration file, depending on your application and network requirements.

> Edge-LB pool has two sub-components:

Reviewer: edge-lb pool is a group of pool server, each pool server has two components that you mentioned below.

Author: Created a separate ticket

- Load balancing manager (lbmgr)
- HAProxy

## Load balancing manager (lbmgr)

> The load balancing manager (`lbmgr`) in Edge-LB pool instances is a non-distributed process that runs along with the HAProxy load balancer. `lbmgr` takes the appropriate action for the pool when it receives requests from the API server.

Reviewer: As I already mentioned, even within our team we lack consistent naming for this component. Some say `pool server`, some say `pool instance`, some call the whole container using the supervisor's process name `lbmgr`. 

Do you have any suggestions regarding the naming? Is `pool server` something you like or would you prefer `pool instance`? What do you think?

Author: Created separate ticket

> The `lbmgr` component monitors the HAProxy instances for the Edge-LB pool and restarts HAProxy in the same container, when needed. For example, if an HAProxy instance goes down because an agent is disconnected, the `lbmgr` process creates a new HAProxy instance to replace it on another available agent.

Reviewer: > it would be the EdgeLB server (not lbmgr) that would try to create a new instance on another agent.

Just to make already great comment from @cneth more precise, it is the SDK scheduler that restart pool server instances in case when they fail. lbmgr restarts the haproxy if needed, but it roles is limited to the single container.

> The `lbmgr` component monitors the HAProxy instances for the Edge-LB pool and restarts HAProxy in the same container, when needed. For example, if an HAProxy instance goes down because an agent is disconnected, the `lbmgr` process creates a new HAProxy instance to replace it on another available agent.

Author: I think I understand, but probably best to check, since I am still new. Is this statement correct?:

"The `lbmgr` component monitors the HAProxy instances for the Edge-LB pool and restarts HAProxy in a single container, when needed. If an HAProxy instance goes down because an agent is disconnected, the SDK scheduler restarts the HAProxy instance on another available agent."

> The `lbmgr` component also validates new load balancer configuration settings and reloads HAProxy load balancer instances, when necessary. For example, the `lbmgr` component manages the timeout interval for reloading the HAProxy instance.

Reviewer: ```suggestion
The `lbmgr` component also validates new HaProxy configuration settings and reloads it when necessary, manages reload timeouts, etc...
```

Author: Changed as noted.
